driver/src/main/java/org/neo4j/driver/internal/LeakLoggingNetworkSession.java

```diff
@@ -49,7 +49,7 @@ private void logLeakIfNeeded()
         if ( isOpen )
         {
             logger.error( "Neo4j Session object leaked, please ensure that your application" +
-                          "calls the `close` method on Sessions before disposing of the objects.\n" +
+                          "fully consumes results in Sessions or explicitly calls `close` on Sessions before disposing of the objects.\n" +
```

Comment:
Does java driver releases connection when the result consumption is complete? Having looked shortly, it seems that `NetworkSession#closeAsync` does the actual release work and it's not called from within the driver at all. Maybe worth checking?

And besides, maybe add ` ` (space) just before `fully`?

Reply:
Both `SessionPullAllHandler#success` (also `failure`) and `Session#closeAsync` will release connection. The `DirectConnection#release` will ensure we release the one and only connection once. If the result is fully consumed, the `SessionPullAllHandler` will first release the connection.